wallet: group outputs only once, decouple it from Coin Selection

furszy opened the pull request:
The idea originates from https://github.com/bitcoin/bitcoin/pull/24845#issuecomment-1130310321.

Note:
For clarity, it's recommended to start reviewing from the end result to understand the structure of the flow.

#### GroupOutputs function rationale:
If "Avoid Partial Spends" is enabled, the function gathers outputs with the same script together inside a container. So Coin Selection can treats them as if them were just one possible input and either select them all or not select them. 

#### How the Inputs Fetch + Selection process roughly works:

```
1. Fetch user’s manually selected inputs.
2. Fetch wallet available coins (walks through the entire wallet txes map) and insert them into a set of vectors (each vector store outputs from a single type).
3. Coin Selection Process:
   Call `AttemptSelection` 8 times. Each of them expands the coin eligibility filter (accepting a larger subset of coins in the calculation) until it founds a solutions or completely fails if no solutions gets founds after the 8 rounds.

   Each `AttemptSelection` call performs the following actions:
     - For each output type supported by the wallet (P2SH, P2PK, P2WPKH, P2WSH and a combination of all of them):
       Call ‘ChooseSelectionResult’ providing the respective, filtered by type, coins vector. Which:
           I. Groups the outputs vector twice (one for positive only and a second one who includes the negative ones as well).
              - GroupOutputs walks-through the entire inputted coins vector one time at least, + more if we are avoiding partial spends, to generate a vector of OutputGroups.
           II. Then performs every coin selection algorithm using the recently created vector of OutputGroup: (1) BnB, (2) knapsack and (3) SRD.
           III. Then returns the best solution out of them.
```

We perform the general operation of gathering outputs, with the same script, into a single container inside:
Each coins selection attempt (8 times —> each coin eligibility filter), for each of the outputs vector who were filtered by type (plus another one joining all the outputs as well if needed), twice (one for the positive only outputs effective value and a second one for all of them).

So, in the worst case scenario where no solution is found after the 8 Coin Selection attempts, the `GroupOutputs` function is called 80 times (8 * 5 * 2).

#### Improvements:

This proposal streamlines the process so that the output groups, filtered by coin eligibility and type, are created in a single loop outside of the Coin Selection Process. 

The new process is as follows:

```
1. Fetch user’s manually selected inputs.
2. Fetch wallet available coins.
3. Group outputs by each coin eligibility filter and each different output type found.
4. Coin Selection Process: 
   Call AttemptSelection 8 times. Each of them expands the coin eligibility filter (accepting different output groups) until it founds a solutions or completely fails if no solutions gets founds after the 8 rounds.
   
   Each ‘AttemptSelection’ call performs the following actions:
      - For each output type supported by the wallet (P2SH, P2PK, P2WPKH, P2WSH and all of them):
          A. Call ‘ChooseSelectionResult’ providing the respective, filtered by type, output group. Which:
             I. Performs every coin selection algorithm using the provided vector of OutputGroup: (1) BnB, (2) knapsack and (3) SRD.
             II. Then returns the best solution out of them.
```

Extra Note:
The next steps after this PR will be to:
1) Merge `AvailableCoins` and `GroupOutputs` processes.
2) Skip entire coin selection rounds if no new coins are added into the subsequent round.
3) Remove global feerates from the OutputGroup class.
4) Remove secondary "grouped" tx creation from `CreateTransactionInternal` by running Coin Selection results over the aps grouped outputs vs non-aps ones.

furszy committed:
wallet: make OutputGroup "positive_only" filter explicit

And not hide it inside the `OutputGroup::Insert` method.
This method does not return anything if insertion fails.

We can know before calling `Insert` whether the coin
will be accepted or not.

furszy committed:
test: wallet, add coverage for outputs grouping process

The following scenarios are covered:

1) 10 UTXO with the same script:
   partial spends is enabled --> outputs must not be grouped.

2) 10 UTXO with the same script:
   partial spends disabled --> outputs must be grouped.

3) 20 UTXO, 10 one from scriptA + 10 from scriptB:
   a) if partial spends is enabled --> outputs must not be grouped.
   b) if partial spends is not enabled --> 2 output groups expected (one per script).

3) Try to add a negative output (value - fee < 0):
   a) if "positive_only" is enabled --> negative output must be skipped.
   b) if "positive_only" is disabled --> negative output must be added.

4) Try to add a non-eligible UTXO (due not fulfilling the min depth target for
 "not mine" UTXOs) --> it must not be added to any group

5) Try to add a non-eligible UTXO (due not fulfilling the min depth target for
 "mine" UTXOs) --> it must not be added to any group

6) Surpass the 'OUTPUT_GROUP_MAX_ENTRIES' size and verify that a second partial
group gets created.

furszy committed:
refactor: make OutputGroup::m_outputs field a vector of shared_ptr

Initial steps towards sharing COutput instances across all possible
OutputGroups (instead of copying them time after time).

furszy committed:
wallet: decouple outputs grouping process from each ChooseSelectionResult

Another step towards the single OutputGroups calculation goal

furszy committed:
wallet: unify outputs grouping process

The 'GroupOutputs()' function performs the same
calculations for only-positive and mixed groups,
the only difference is that when we look for
only-positive groups, we discard negative utxos.

So, instead of wasting resources calling GroupOutputs()
for positive-only first, then call it again to include
the negative ones in the result, we can execute
GroupOutputs() only once, including in the response
both group types (positive-only and mixed).

furszy committed:
wallet: single output groups filtering and grouping process

Optimizes coin selection by performing the "group outputs"
procedure only once, outside the "attempt selection" process.

Avoiding the repeated execution of the 'GroupOutputs' operation
that occurs on each coin eligibility filters (up to 8 of them);
then for every coin vector type plus one for all the coins together.

This also let us not perform coin selection over coin eligibility
filtered groups that don't add new elements.
(because, if the previous round failed, and the subsequent one has
the same coins, then this new round will fail again).

aureleoules commented on src/wallet/test/group_outputs_tests.cpp:
```diff
@@ -0,0 +1,230 @@
+{
+    FastRandomContext rand{};
+    return CoinSelectionParams{
+            rand,
```
`CoinSelectionParams`'s `rng_fast` member is a reference and you're using a stack object, this will create a dangling reference.

furszy replied:
oops, good catch 👍🏼

aureleoules commented on src/wallet/coinselection.h:
```diff
@@ -229,11 +235,41 @@ struct OutputGroup
+    // All groups, no filters
+    Groups all_groups;
+
+    enum InsertGroupType {
```
prefer enum class

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -416,28 +416,35 @@ std::map<CTxDestination, std::vector<COutput>> ListCoins(const CWallet& wallet)
-            // Check the OutputGroup's eligibility. Only add the eligible ones.
-            if (positive_only && group.GetSelectionAmount() <= 0) continue;
-            if (group.m_outputs.size() > 0 && group.EligibleForSpending(filter)) groups_out.push_back(group);
+        for (const auto& coins_by_type : coins.coins) {
```
could use structured binding

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -474,39 +477,69 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
-    for (const auto& spk_and_groups_pair: spk_to_groups_map) {
-        const std::vector<OutputGroup>& groups_per_spk= spk_and_groups_pair.second;
+    // Now we go through the entire maps and pull out the OutputGroups
+    const auto& push_output_groups = [&](ScriptPubKeyToOutgroup& groups_map ,bool positive_only) {
```
```suggestion
    const auto& push_output_groups = [&](const ScriptPubKeyToOutgroup& groups_map ,bool positive_only) {
```

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -474,39 +477,69 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
-        const std::vector<OutputGroup>& groups_per_spk= spk_and_groups_pair.second;
+    // Now we go through the entire maps and pull out the OutputGroups
+    const auto& push_output_groups = [&](ScriptPubKeyToOutgroup& groups_map ,bool positive_only) {
+        for (const auto& spk_and_groups_pair: groups_map) {
```
could use structured binding

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -517,40 +550,39 @@ std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAm
 };
 
-std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const std::vector<COutput>& available_coins, const CoinSelectionParams& coin_selection_params)
+util::Result<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const CAmount& nTargetValue, Groups& groups,
```
wallet is unused

furszy replied:
good eye, this let me remove the `AttemptSelection` wallet arg too.

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -614,54 +646,54 @@ std::optional<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coi
-                                   available_coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) {
-                return r5;
-            }
+            ordered_filters.emplace_back(CoinEligibilityFilter(0, 1, 2));
```
No need for temporary object, same around this line.
```suggestion
            ordered_filters.emplace_back(0, 1, 2);
```

furszy replied:
I actually didn't make this change on purpose to minimize the diff burden inside the same commit. Would prefer to do it in a different commit to make reviewers' life a bit easier (adding the args names there too).

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -925,7 +958,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     const uint32_t nSequence{coin_control.m_signal_bip125_rbf.value_or(wallet.m_signal_rbf) ? MAX_BIP125_RBF_SEQUENCE : CTxIn::MAX_SEQUENCE_NONFINAL};
     for (const auto& coin : selected_coins) {
-        txNew.vin.push_back(CTxIn(coin.outpoint, CScript(), nSequence));
+        txNew.vin.push_back(CTxIn(coin->outpoint, CScript(), nSequence));
```
could use emplace_back
```suggestion
        txNew.vin.emplace_back(coin->outpoint, CScript(), nSequence);
```

furszy replied:
agree but it's unrelated to the commit changes. Let's better leave it for another PR.

aureleoules commented on src/wallet/test/group_outputs_tests.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    const uint256& txid = tx.GetHash();
+    LOCK(wallet.cs_wallet);
+    auto ret = wallet.mapWallet.emplace(std::piecewise_construct, std::forward_as_tuple(txid), std::forward_as_tuple(MakeTransactionRef(std::move(tx)), TxStateInactive{}));
```
could use structured binding

fanquake commented:
PR description needs be updated to remove the "this is built on", for the already-merged PR. Although, given `Footnote; this PR is still a WIP:` in the PR description, maybe this should also be a draft (that would better communicate the WIP intent that a footnote in description)? If it's no-longer WIP, please update the PR description to remove that as well. Can be done along with the next rebase.

furszy commented:
> PR description needs be updated to remove the "this is built on", for the already-merged PR. 

removed, thanks.

> Although, given `Footnote; this PR is still a WIP:` in the PR description, maybe this should also be a draft (that would better communicate the WIP intent than a footnote in description)? If it's no-longer WIP, please update the PR description to remove that as well. Can be done along with the next rebase.

yeah. This one is on my current working-path. I'm planning to update the PR description soon.
Since yesterday, have been working on https://github.com/bitcoin/bitcoin/pull/26560#discussion_r1038307694 which will make this PR a bit smaller.

furszy committed:
test: coinselector_tests refactor, use CoinsResult instead of plain std::vector

No functional changes. Only cosmetic changes to simplify the follow-up commit.

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -536,31 +539,33 @@ util::Result<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmo
+        const auto& all = available_coins.All();
+        Groups groups;
+        groups.positive_group = GroupOutputs(wallet, all, coin_selection_params, eligibility_filter, true /* positive_only */);
+        groups.mixed_group = GroupOutputs(wallet, all, coin_selection_params, eligibility_filter, false /* positive_only */);
```
nit: same as above (comment before value)

furszy replied:
same answer as above :)

aureleoules reviewed (commented):
Left some minor comments. Will review 6df2b735b6ad1e744b2ecceeee10fb36a6673fd6, 5a7ad35fb682446efb55f93b2f2f07eadd8fc0e3, and 325f171145771e5ff3a4c7e7eb0966728709370b further later.

### Benchmark results
`./src/bench/bench_bitcoin -filter='CoinSelection' --min-time=5000`.

#### Master

|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|          426,716.89 |            2,343.47 |    0.3% |   10,501,652.22 |   2,344,328.92 |    0.2% |      5.41 | `CoinSelection`

#### PR :rocket: 

|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|          328,320.35 |            3,045.81 |    0.2% |    8,324,339.49 |   1,946,693.33 |    0.1% |      5.34 | `CoinSelection`

I ran the benchmarks multiple times to make sure they are consistent.

furszy commented on src/wallet/test/group_outputs_tests.cpp:
```diff
@@ -0,0 +1,224 @@
+
+static int nextLockTime = 0;
+
+static std::shared_ptr<CWallet> NewWallet(const node::NodeContext& m_node, const ArgsManager& m_args)
+{
+    std::unique_ptr<CWallet> wallet = std::make_unique<CWallet>(m_node.chain.get(), "", m_args, CreateMockWalletDatabase());
+    wallet->LoadWallet();
+    LOCK(wallet->cs_wallet);
+    wallet->SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
+    wallet->SetupDescriptorScriptPubKeyMans();
+    return wallet;
+}
```
Great catch!
Cleaned inside https://github.com/bitcoin/bitcoin/pull/26720#issuecomment-1397832543.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -521,8 +521,11 @@ util::Result<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmo
+    for (const auto& [type, coins] : available_coins.coins) {
+        Groups groups;
+        groups.positive_group = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, true /* positive_only */);
+        groups.mixed_group = GroupOutputs(wallet, coins, coin_selection_params, eligibility_filter, false /* positive_only */);
```
This lines were just moved from one place to another (tried to not change them to make the review easier), and are later removed inside [b187c3a](https://github.com/bitcoin/bitcoin/pull/25806/commits/b187c3a7e79aa8768196353a809a4d03996e44a4).

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -375,17 +372,53 @@ CAmount OutputGroup::GetSelectionAmount() const
-        selected_effective_value += use_effective_value ? coin.GetEffectiveValue() : coin.txout.nValue;
+    for (const auto& coin : inputs) {
+        waste += coin->GetFee() - coin->long_term_fee;
+        selected_effective_value += use_effective_value ? coin->GetEffectiveValue() : coin->txout.nValue;
```
done

theStack commented on src/wallet/test/group_outputs_tests.cpp:
```diff
@@ -0,0 +1,224 @@
+
+    const CTxDestination dest4 = *Assert(wallet->GetNewDestination(OutputType::BECH32, ""));
+    addCoin(group_verifier.coins_pool, *wallet, dest4, 6 * COIN,
+            /*is_from_me=*/false, CFeeRate(0), /*depth=*/2);
```
nit: could use the just-below-boundary value of 5 here
```suggestion
            /*is_from_me=*/false, CFeeRate(0), /*depth=*/5);
```

furszy replied:
done

theStack commented on src/wallet/test/group_outputs_tests.cpp:
```diff
@@ -0,0 +1,224 @@
+    // ###########################################################################################
+
+    const CTxDestination dest7 = *Assert(wallet->GetNewDestination(OutputType::BECH32, ""));
+    for (unsigned long i = 0; i < 130; i++) { // OUTPUT_GROUP_MAX_ENTRIES{100}
```
nit: could just-above-boundary value of 101 here
```suggestion
    for (unsigned long i = 0; i < 101; i++) { // OUTPUT_GROUP_MAX_ENTRIES{100}
```

furszy replied:
done

theStack reviewed (commented):
Concept ACK

theStack commented on src/wallet/spend.h:
```diff
@@ -132,18 +132,15 @@ util::Result<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmo
  */
-util::Result<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const std::vector<COutput>& available_coins,
-                        const CoinSelectionParams& coin_selection_params);
+util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue, Groups groups, const CoinSelectionParams& coin_selection_params);
```
nit: at this point one could already pass the `groups` parameter by-reference rather than by-value (it is done later in eca918dd17739036cbb2430a4fc2a5a785d88fb6 though anyways) : 
```suggestion
util::Result<SelectionResult> ChooseSelectionResult(const CAmount& nTargetValue, Groups& groups, const CoinSelectionParams& coin_selection_params);
```

furszy replied:
sure, will re-order it so the change appears when `Groups` are introduced and not later.

theStack commented on src/wallet/coinselection.h:
```diff
@@ -247,6 +248,27 @@ struct Groups {
+        BOTH // Insert to `positive_group` and `mixed_group`
+    };
+
+    // Depending on the 'insert_type', appends the output group into the 'by type' groups
+    // and/or the 'all groups' container.
```
IIUC, for all possible `insert_types`, something is inserted both in the `by type` groups and `all groups` container. The `and/or` seems to apply for the `positive_group`/`mixed_group`? Suggestion:
```suggestion
    // Depending on the 'insert_type', appends the output group into the `positive_group`
    // and/or the 'mixed_group' for both the `by type` groups and the `all groups` container.
```

furszy replied:
Yeah, thanks!.
Aside from the method doc re-phrasing, have re-worked the area, so it's clear what the `push` method does on each scenario.

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -372,6 +372,42 @@ CAmount OutputGroup::GetSelectionAmount() const
     return m_subtract_fee_outputs ? m_value : effective_value;
 }
 
+void OutputGroups::push(const OutputGroup& group, OutputType type, InsertGroupType insert_type)
```
pushed

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -372,6 +372,42 @@ CAmount OutputGroup::GetSelectionAmount() const
+    }
+}
+
+std::optional<Groups> OutputGroups::find(OutputType type)
```
pushed

theStack reviewed (commented):
I was thinking if the `OutputGroups` class could be simplified somehow, right now the methods `Push` and `PushAll` seem to be a bit redundant and could be deduplicated. How about a single method like
```
void OutputGroups::Push(const OutputGroup& group, OutputType type, bool insert_positive, bool insert_mixed)
{
    if (group.m_outputs.empty()) return;

    Groups& groups = groups_by_type[type];	
    if (insert_positive && group.GetSelectionAmount() > 0) {
        groups.positive_group.emplace_back(group);
        all_groups.positive_group.emplace_back(group);
    }
    if (insert_mixed) {
        groups.mixed_group.emplace_back(group);
        all_groups.mixed_group.emplace_back(group);
    }
}
```
(This seems a bit more like the earlier version you had, but without enums and without duplication.) The methods `OutputGroups::PushAll` and `Groups::GetRef` could then simply be dropped and only the call-sites need to be adapted w.r.t. to the two boolean parameters.

theStack reviewed (approved):
Code-review ACK 0bb90cabc8e2c6bbbc3c52ec1170e725435b625d

achow101 commented:
ACK 0bb90cabc8e2c6bbbc3c52ec1170e725435b625d

S3RK commented:
Concept ACK. started review

S3RK commented on src/wallet/coinselection.h:
```diff
@@ -237,11 +243,37 @@ struct OutputGroup
     CAmount GetSelectionAmount() const;
 };
 
+struct Groups {
```
I believe the separation between positive and mixed group is an implementation detail of particular coin selection algorithms. We can contrast it with `CoinEligibilityFilter` and `OutputType` grouping which are algorithm independent. Therefore I propose to not separate positive outputs at the grouping stage.

Rather filter negative/positive groups  as close to the particular coin selection algo as possible, e.g. in `ChooseSelectionResult`. This will significantly simplify the code by removing extra nesting of structs and multiple `if` statements. I expect computation overhead of iterating over groups to remove negative UTXos to be negligible, but maybe I'm wrong.

furszy replied:
hmm, that would make the avoid partial spends flow even uglier, plus I don't think that that will be negligible for big wallets (+50k UTXOs wallets).

The later removal of negative UTXOs will leave APS full groups with less than 100 entries, which will force us to re-order groups so they are full again (moving UTXOs from one group to another), then if the wallet does not include partial groups, will need to discard those groups too.
So making this changes would actually introduce more code than what we currently have with few nested structs whose creation is scoped only to `GroupOutputs`. I mean, we would need to have one APS procedure inside `GroupOutputs` and another APS procedure inside `ChooseSelectionResult`.

About the computation overhead:
In the worst case scenario, the wallet executes `ChooseSelectionResult` 5 times in each of the 8 `AttemptSelection` calls.
So, it could end up performing the traverse + re-order + removal 40 times.

S3RK replied:
I'm not sure this is how APS works. When we remove negative UTXOs we assess the whole group and not each UTXO individually, so you don't need to regroup anything after filtering negative groups.

furszy replied:
> I'm not sure this is how APS works. When we remove negative UTXOs we assess the whole group and not each UTXO individually, so you don't need to regroup anything after filtering negative groups.

No wait, that would leave negative UTXOs in positive-only groups if the sum of the positive UTXOs surpasses the sum of the negative ones.

Currently, a "positive-only group" is essentially a container who stores only positive UTXOs. Not a mix of UTXOs which values sum is positive. That is different conceptually and something that, at least for now, we are not allowing.

Check master, inside `OutputGroup::Insert` method, first line. We reject individual negative UTXOs when the container only stores positive-only UTXOs.

The only way to create positive-groups out of mixed groups is to traverse each group UTXOs set and remove the negative ones. Then, if the group has less than a 100 entries and there are other groups corresponding to the same script (after doing the same removal on those groups), the wallet need to re-order them so only full groups are passed to the selection algorithms (if the avoid partial groups flag is enabled).

Aside from that, performance wise, I still think that performing the negative UTXOs removal later on the process will introduce a non-negligible overhead in big wallets.

S3RK replied:
You're right, thanks for the explanation. This comment could be resolved.

murchandamus reviewed (commented):
Concept ACK, very light review, need to go over it in more detail still

furszy commented on src/wallet/spend.cpp:
```diff
@@ -417,13 +417,14 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
-            OutputGroup group{coin_sel_params};
-            group.Insert(output, ancestors, descendants, positive_only);
+            // If 'positive_only' is set, filter for positive only before adding the coin
+            if (!positive_only || (positive_only && output.GetEffectiveValue() > 0)) {
```
yeah k 👌🏼, will do the change if need to re-touch to not invalidate the other reviews only for this small redundancy.

S3RK commented on src/wallet/coinselection.h:
```diff
@@ -237,11 +243,37 @@ struct OutputGroup
+    // Stores 'OutputGroup' containing only positive UTXOs (value > 0).
+    std::vector<OutputGroup> positive_group;
+    // Stores 'OutputGroup' which may contain both positive and negative UTXOs.
+    std::vector<OutputGroup> mixed_group;
```
nit: it'd be great if we can disambiguate the term `mixed`. Now it applies to both a) mixed type groups b) mixed effects value group

furszy replied:
What are you referring with "mixed type groups"?

The `mixed` term is only used inside the `Groups` structure, and reflects a vector of `OutputGroup` that may contain positive and negative UTXOs.

furszy replied:
Ok understood it, the `mixed` term is also used for the `allow_mixed_output_types` flag. Which is not from this PR.

S3RK commented on src/wallet/coinselection.h:
```diff
@@ -237,11 +243,37 @@ struct OutputGroup
+    // Map filters to output groups.
+    std::map<OutputType, Groups> groups_by_type;
+    // All groups, no filters
+    Groups all_groups;
```
nit: `All` groups is somewhat misleading here, because these are actually filtered groups, but for all output types. Unfortunately, I don't have better suggestion yet.

furszy replied:
The "filter" term usage here in the comment is wrong. For what corresponds to this struct, the `all` term just refers to the combination of all inserted groups. Same as we do in the `CoinsResult::All` method. The filtering process, nor what is inserted (whether is filtered or not), is not responsibility of this struct.

It's just a container that keeps track of `Groups` by type and the caches the combination of all of them.

furszy replied:
I mean, it would be slower but could also be created on-demand:
```c++
Groups OutputGroups::All()
{
    Groups ret;
    for (const auto& [type, groups_vec] : groups_by_type) {
        ret.positive_group.insert(ret.positive_group.begin(), groups_vec.positive_group.begin(), groups_vec.positive_group.end());
        ret.mixed_group.insert(ret.mixed_group.begin(), groups_vec.mixed_group.begin(), groups_vec.mixed_group.end());
    }
    return ret;
}
```

furszy replied:
Fixed the "filter" comment mention.

S3RK commented on src/wallet/coinselection.h:
```diff
@@ -237,11 +243,37 @@ struct OutputGroup
     CAmount GetSelectionAmount() const;
 };
 
+struct Groups {
```
nit: Judging from the name it's not clear what's the difference between `Groups` and `OutputGroups`

furszy replied:
`GroupsByValueSign`? 

I prefer the struct rather than having an ugly std::pair or a typedef:
`typedef std::pair<std::vector<OutputGroup>, std::vector<OutputGroup>> GroupsByValueSign`

furszy replied:
Ended up renaming `OutputGroups` to `OutputGroupTypeMap` to reflect what the struct really does and differentiate it from the inner struct.

S3RK commented:
Code review ACK 0bb90cabc8e2c6bbbc3c52ec1170e725435b625d modulo tests.

I believe this PR is a good improvement over status quo.

The biggest worry I have now it somewhat ambiguous terminology:
- `mixed` groups could mean mixed type or mixed effective value
-  `all` groups not always refers to actually all groups, but filtered by `CoinEligibiltyFilter`
- two structs `Groups` vs `OutputGroups`

I don't have better suggestion as of now, so it's not blocking. I'll think more and come back if I have some ideas. Maybe @Xekyo the master of terminology have some ideas.

furszy commented on src/wallet/coinselection.cpp:
```diff
@@ -372,6 +372,28 @@ CAmount OutputGroup::GetSelectionAmount() const
     return m_subtract_fee_outputs ? m_value : effective_value;
 }
 
+void OutputGroups::Push(const OutputGroup& group, OutputType type, bool insert_positive, bool insert_mixed)
```
I wrote it in that way before but changed it due a previous code review suggestion (previous [diff](https://github.com/bitcoin/bitcoin/compare/4343835a23290f2cd34404821494303cc75ce86e..5246cc1098e5409e986c1f215bc88eb44cec62f8)). I'm not fan of bool args neither but, at least here, I would be ok either way.

S3RK replied:
The reason why I suggested this change is because it's hard to explain why we need a `insert_mixed` flag. Why don't we just always add groups to `mixed` vector? I understand that's because of how aps works, but just looking at this method it's mysterious.

furszy replied:
well, this method is just a push function. It doesn't have to tell anything about the context where it is used. The struct is basically a map container with a cache, it doesn't need to explain how APS works.

If you are strong over the bool args usage, that would be the argument to make the changes for me.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -467,28 +467,49 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+    for (const auto& [type, outs] : coins.coins) {
+        for (const COutput& output : outs) {
+            // Skip outputs we cannot spend
+            if (!output.spendable) continue;
```
yes, I didn't remove it just to make the diff smaller. This was already there.

furszy commented on src/wallet/coinselection.h:
```diff
@@ -193,6 +193,11 @@ struct CoinEligibilityFilter
     CoinEligibilityFilter(int conf_mine, int conf_theirs, uint64_t max_ancestors, uint64_t max_descendants) : conf_mine(conf_mine), conf_theirs(conf_theirs), max_ancestors(max_ancestors), max_descendants(max_descendants) {}
     CoinEligibilityFilter(int conf_mine, int conf_theirs, uint64_t max_ancestors, uint64_t max_descendants, bool include_partial) : conf_mine(conf_mine), conf_theirs(conf_theirs), max_ancestors(max_ancestors), max_descendants(max_descendants), m_include_partial_groups(include_partial) {}
+
+    bool operator<(const CoinEligibilityFilter& other) const {
```
for the `FilteredOutputGroups` map. `std::map` requires a comparator or a hash function.

furszy commented on src/wallet/coinselection.h:
```diff
@@ -249,6 +250,21 @@ struct Groups {
+/** Stores several 'Groups' whose were mapped by output type. */
+struct OutputGroups
+{
+    // Map filters to output groups.
```
yeah

furszy commented on src/wallet/spend.h:
```diff
@@ -106,7 +106,14 @@ const CTxOut& FindNonChangeParentOutput(const CWallet& wallet, const COutPoint&
+/**
+* Group coins by the provided filters.
+*/
+OutputGroups GroupOutputs(const CWallet& wallet,
```
yeah could do that, but we wouldn't win much really. The commit that touches
this method signature is the biggest one of the PR, so we would add more diff
for no real gain.

Better to leave it for the next PR on the line. I have few that will come right after
this one anyway.

furszy commented:
Thanks @S3RK for the review.

Updated per feedback ([small diff](https://github.com/bitcoin/bitcoin/compare/0bb90cabc8e2c6bbbc3c52ec1170e725435b625d..ae0dd2becd4fa01666793c3f48e8725d5e585c30)), tackled the following points that made sense for me to include here:

* Renamed `OutputGroups` to `OutputGroupTypeMap` to reflect what the struct really does and differentiate it from the inner struct.
* Corrected the `OutputGroups`, now `OutputGroupTypeMap`, members comments.
* And tackled @Xekyo’s unneeded check.

I think that while we all agree that this work is functionality and structurally sound, we can move forward.
And tackle any new naming discussion in a follow-up work.

furszy commented:
Rebased it due silent merge conflicts with #26889.
Ready to go now.

S3RK commented:
utACK 1cfed6ce6cd8c9f9f1e93662ff8f92cd5a5b91c4 (also reviewed the tests)

furszy commented on src/wallet/test/group_outputs_tests.cpp:
```diff
@@ -0,0 +1,225 @@
+        addCoin(group_verifier.coins_pool, *wallet, dest7, 9 * COIN, /*is_from_me=*/true);
+    }
+
+    // Exclude partial groups must have no changes from this round and the previous one (point 6)
```
sounds good

furszy commented:
Updated per feedback, thanks S3RK 👌🏼.

[Small diff](https://github.com/bitcoin/bitcoin/compare/1cfed6ce6cd8c9f9f1e93662ff8f92cd5a5b91c4..6a302d40dfcc33b15bf27a1d3fcc04bfd3cbda44), changes are in the last test scenario and the test commit message.

S3RK commented:
ReACK 6a302d4
Thanks for addressing feedback

theStack reviewed (approved):
re-ACK 6a302d40dfcc33b15bf27a1d3fcc04bfd3cbda44 :coconut: 

(verified via `git range-diff 0bb90cab...6a302d40` that since my last ACK, most review comments by Murch and S3RK have been tackled, mostly trivial refactor + terminology/naming improvements and [a minor unit test adaption](https://github.com/bitcoin/bitcoin/pull/25806#discussion_r1126088720))

achow101 commented:
ACK 6a302d40dfcc33b15bf27a1d3fcc04bfd3cbda44

murchandamus reviewed (commented):
Post merge ACK 6a302d40dfcc33b15bf27a1d3fcc04bfd3cbda44
